### PR TITLE
os versions, os download: Add support for balenaOS ESR versions

### PR DESCRIPTION
**Pull request by pdcastro**
Resolves: #2005
Depends-on: balena-io-modules/balena-image-manager/pull/54
See: https://www.flowdock.com/app/rulemotion/resin-devices/threads/BtGUlpKl9fyhDGpI1OJ4W32EfAD
Change-type: minor

### CODE REVIEW

File: lib/utils/cloud.ts
Comment (pdcastro) [inline lib/utils/cloud.ts:158]: Is it always correct to say `balenaOS` here, instead of `Device OS`?
Sample output after these changes:
```
Downloading balenaOS version 2.83.21+rev1.prod [========                ] 34% eta 30s
```
Whereas previously it might have read:
```
Downloading Device OS recommended [=                       ] 3% eta 1m17s
```

File: lib/utils/cloud.ts
Comment (pdcastro) [inline lib/utils/cloud.ts:153]: This relies on a new `'version'` event introduced in balena-io-modules/balena-image-manager/pull/54

**Comment by shawaj** [comment]
Any idea when this will be available?

We need to download and preload ESR images

**Comment by pdcastro** [comment]
> Any idea when this will be available?
> We need to download and preload ESR images

@shawaj, this PR depends on balena-io-modules/balena-image-manager/pull/54 , which is waiting on a code review. A colleague said they'll try to review it tomorrow. Meanwhile, are you able to download the ESR image through the web dashboard, and then preload it?

File: lib/commands/os/versions.ts
Comment (thgreasi) [inline lib/commands/os/versions.ts:59]: If you bump to the latest, you can instead use `getAvailableOsVersions()`, since after the next major, `getAllOsVersions()` is also going to include invalidated versions as well.

**Review by thgreasi — APPROVED** [review]
:+1:

File: lib/utils/cloud.ts
Comment (pdcastro) [reply lib/utils/cloud.ts:242]: Initially I thought of doing that, but then I realised that the `formattedVersion` field for ESR versions had "richer details" than just "recommended":
```
$ ./bin/balena os download --version menu-esr -o os.img raspberrypi3
Getting device operating system for raspberrypi3
? Select the OS version: (Use arrow keys)
❯ v2021.10.2.prod (next, recommended)
  v2021.10.2.dev (next)
  v2021.07.1.prod (current)
  v2021.07.1.dev (current)
  v2021.04.0.prod (sunset)
  v2021.04.0.dev (sunset)
  v2021.01.0.prod (outdated)
(Move up and down to reveal more choices)
```
Which I thought could good for the interactive menu. If we actually wanted to hide/delete information such as "next", "current" and "sunset", then yes we could just check the `isRecommended` attribute.

The formatting code is now also being shared between `os download` and `os versions`:

```
$ ./bin/balena os versions --esr raspberrypi3
v2021.10.2.prod (next, recommended)
v2021.10.2.dev (next)
v2021.07.1.prod (current)
v2021.07.1.dev (current)
v2021.04.0.prod (sunset)
v2021.04.0.dev (sunset)
v2021.01.0.prod (outdated)
v2021.01.0.dev (outdated)
v2020.07.2.prod (outdated)
v2020.07.2.dev (outdated)
v2020.04.0.prod (outdated)
v2020.04.0.dev (outdated)
v2019.10.2.prod (outdated)
v2019.10.2.dev (outdated)
```

File: lib/commands/os/versions.ts
Comment (pdcastro) [inline lib/commands/os/versions.ts:60]: In my first go at this PR, I did not have a `--esr` flag and all versions were listed together. But some issues mentioned for `os download` also applied here:

* Two versions labeled "recommended", one ESR and one non-ESR. We could amend the formatting, but:
  * If we didn't print "recommended" at all, it would strictly be a breaking change (re output) for non-ESR versions.
  * If we printed "recommended" only for non-ESR versions, then... are we implying that ESR versions are not recommended? :-)
  * If we did some other special formatting thing for `os versions`, then the formatting code could not be shared with the interactive menu of `os download`.

* Re sorting, having ESR first and non ESR later or vice versa, there was still an element of the issue mentioned in a comment for the `os download` command: whatever we list at the top _with a recommended label_ is not necessarily the best (actually recommended) option for the user. With a `--esr` flag, the "recommendation" always fits the user's choice of ESR or non ESR.

* And finally, the web dashboard has a separate dropdown box to select ESR versions, avoiding listing them together. That solution has a 1-1 mapping to a `--esr` flag, which allows me to claim "platform UI/UX consistency" as well. :-)

File: lib/commands/os/versions.ts
Comment (pdcastro) [reply lib/commands/os/versions.ts]: Done 👍

File: lib/utils/cloud.ts
Comment (pdcastro) [reply lib/utils/cloud.ts]: Done 👍  (A version bump wasn't needed, SDK version `15.51.1` includes `getAvailableOsVersions`)

**Comment by pdcastro** [comment]
@balena-ci I self-certify!

**Comment by pdcastro** [comment]
@thgreasi, only after this PR was merged, I noticed that `sdk.models.hostapp.getAvailableOsVersions()` does not accept device type aliases like `'nuc'`, whereas `sdk.models.os.getSupportedVersions()` accepts aliases. As a result, the `balena os versions` command no longer accepts aliases. I seem to remember a previous conversation regarding aliases support for the hostapp model -- is this a thing? What would you suggest we did for `os versions` and `os download` to support device type aliases?

File: lib/commands/os/versions.ts
Comment (thgreasi) [reply lib/commands/os/versions.ts:63]: @pdcastro it probably surfaces an issue with the original hostApp release migration/import. The SDK & CLI are doing the correct thing. Given how old those releases are though and our plan to deprecate such old releases, I wouldn't assign time on it unless a users asks about it.

**Comment by thgreasi** [comment]
I'm indeed working on adding the aliases to the model, but I don't have an estimate & I'm also about to release a major of the SDK.
A workaround for the regression is to do `await sdk.models.device.getManifestBySlug('slugOrAlias').slug`.

File: lib/commands/os/download.ts
Comment (dfunckt) [reply lib/commands/os/download.ts:88]: > If users typed any of the combinations below, surely they'd expect the "latest", "recommended" or "default" qualifiers to apply to ESR versions rather than non-ESR versions, even if the help output said that the --esr flag could only be used with --version menu, and even if the CLI printed an error if --esr was used with anything other than --version menu. As it happens, those qualifiers only apply to non-ESR versions as currently implemented in the SDK or API.
>
> - `--esr --version latest`
> - `--esr --version recommended`
> - `--esr --version default`

For all modifiers the version for ESR is the same @pdcastro so there is no conflict really and the `download` command could go with the `--esr` argument as well.

**Pull request merged**
